### PR TITLE
Onedal algorithms backed by nuget packages

### PR DESCRIPTION
**Pull request by rgesteve**
Supercedes #6373 and #6364 allowing for full build of functionality without extra packages (they're downloaded by `build.sh`).  Also includes sample to illustrate use.

We are excited to review your PR.

So we can do the best job, please check:

- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [ ] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [x] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [x] You have included any necessary tests in the same PR.

### CODE REVIEW

**Comment by ericstj** [comment]
@rgsteve @michaelgsharp -- I gave a local test of this and was able to reproduce the problem.  It looks like the error is due to missing `onedal_core.1.dll` which is a dependency of `OneDalNative.dll`.  You can see this if you enable loader snaps in the native debugger.  The .NET runtime can't distinguish between a failure from the library itself vs some dependency as the failure happens inside the call to loadlibrary and the OS doesn't distinguish.

File: test/Microsoft.ML.Tests/TrainerEstimators/OneDalEstimators.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+        //[Fact]
+        //[NativeDependencyFact("MatrixFactorizationNative")]
+        [NativeDependencyFact("OneDalNative")]
+        public void TestDependency()
+        {
+            var currentDir = AppContext.BaseDirectory;
+            Output.WriteLine($"**** Running from directory {currentDir}.");
+
+            var dllDir = AppContext.GetData("NATIVE_DLL_SEARCH_DIRECTORIES").ToString();
```
Comment (ericstj) [inline test/Microsoft.ML.Tests/TrainerEstimators/OneDalEstimators.cs:91]: This doesn't work on .NETFramework, but I image you've just added this for debugging purposes.  Make sure to remove it.

**Comment by ericstj** [comment]
I tried to manually workaround the missing `onedal_core.1.dll` by finding it in the inteldal.redist.win-x64 nuget package and manually copying it over, then the test hit errors for other missing dlls:
```
Intel oneDAL FATAL ERROR: Cannot find/load library onedal_thread.1.dll.
Intel oneDAL FATAL ERROR: Cannot find/load library onedal_sequential.1.dll.
Intel oneDAL FATAL ERROR: Cannot load neither onedal_thread.1.dll nor onedal_sequential.1.dll
```
I couldn't find those so it seems blocked.

**Comment by rgesteve** [comment]
Thank you Eric!  Loading OneDalNative successfully and failing to load the dependencies was my observation as well in Linux:
<pre>
 193059:[pid 21092] openat(AT_FDCWD, "/data/projects/machinelearning/.dotnet/shared/Microsoft.NETCore.App/6.0.9/libOneDalNative.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
 193060:[pid 21092] openat(AT_FDCWD, "/data/projects/machinelearning/artifacts/bin/Microsoft.ML.Tests/Debug/net6.0/libOneDalNative.so", O_RDONLY|O_CLOEXEC) = 415
 193069:[pid 21092] openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 415
 193073:[pid 21092] openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libonedal_core.so.1.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
 193074:[pid 21092] openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/libonedal_core.so.1.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
 193075:[pid 21092] openat(AT_FDCWD, "/lib/libonedal_core.so.1.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
</pre>
Like you said, the loader doesn't even bother trying to find the dependencies in directories other than the system-wide ldconfig standard ones.   `ldd` does show `libonedal_{core,thread}` as dependencies of `OneDalCore`, but your suggestion of the library dynamically loading its dependencies makes a lot of sense, especially in light that, pointing LD_LIBRARY_PATH to the exact same directory, with the exact same contents that the one the workload should be search works.   The last commit includes the dependencies in the nupkg for distribution.  I don't expect that this will change the loader issue much, but at least it'll allow people to try explictitly pointing `LD_LIBRARY_PATH`/`PATH` to where the dependencies are in their filesystem.

In the meanwhile will work with the oneDAL team to see if we can figure out what's going on.

File: test/Microsoft.ML.Tests/TrainerEstimators/OneDalEstimators.cs (hunk 1 above)
Comment (rgesteve) [reply test/Microsoft.ML.Tests/TrainerEstimators/OneDalEstimators.cs:91]: Will do, this whole test is there on a *very* temporary basis.

**Comment by ericstj** [comment]
It might be worthwhile following up with Interop folks like @AaronRobinsonMSFT and @elinor-fung  to understand if it is expected that the runtime doesn’t set this native probing path itself. It could be something odd about how ML.Net runs tests.

File: test/Microsoft.ML.Tests/TrainerEstimators/OneDalEstimators.cs (hunk 1 above)
Comment (rgesteve) [reply test/Microsoft.ML.Tests/TrainerEstimators/OneDalEstimators.cs]: Commit #e2a60945 removes the test.

**Review by michaelgsharp — APPROVED** [review]
:shipit:

**Pull request merged**